Drop tablets of dropped views and indices

bhalevy opened the pull request:
This series adds notification before dropping views and indices so that the
tablet_allocator can generate mutations to respectively drop all tablets associated with them from system.tablets.

Additional unit tests were added for these cases.

Note that one case is not yet tested: where a table is allowed to be dropped while having views that depend on it, when it is dropped from the alternator path.

This is tested indirectly by testing dropping a table with live secondary index as it follows the same notification path as views in this series.

Fixes #17627

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7347/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7347/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/455/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8488/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7347/)
#### Build Details:
- Duration: 1 hr 54 min
- Builder: spider5.cloudius-systems.com

nyh commented:
Thanks, good catch, I'll review the patch now.
In commits 88a5ddabcec625434b2df0ed82a63c36245af35f and 8181e287312c2b3eefa222eb6946cdae29195aa5 I added the missing notifier calls when creating views and indexes, I should have realized we may have a similar problem when deleting them, but when I saw no failure in the tests I assumed everything was fine :-(

We should have (at least in debug mode) some sort of "leak checker", if tablet data exists that doesn't belong to any table, it's a bug. It would have caught this bug.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -107,3 +107,17 @@ def test_describe_initial_tablets(cql, skip_without_tablets):
+        with new_test_table(cql, keyspace, "pk int PRIMARY KEY, c int") as table:
+            table_name = table.split('.')[1]
+            res = cql.execute(f"SELECT * FROM system.tablets WHERE keyspace_name='{keyspace_name}' AND table_name='{table_name}' ALLOW FILTERING")
+            assert res is not None, f"table {keyspace_name}.{table_name} not found in system.tablets after the table was created"
```
Are you sure this is the right assert? I would have assumed that if there are zero results, you still get some object (not "None") which has the empty list of results?

Note that below you used "not res", not "res is None" - those are not the same thing.

bhalevy replied:
fixed in next version

bhalevy commented:
> Thanks, good catch, I'll review the patch now. In commits [88a5dda](https://github.com/scylladb/scylladb/commit/88a5ddabcec625434b2df0ed82a63c36245af35f) and [8181e28](https://github.com/scylladb/scylladb/commit/8181e287312c2b3eefa222eb6946cdae29195aa5) I added the missing notifier calls when creating views and indexes, I should have realized we may have a similar problem when deleting them, but when I saw no failure in the tests I assumed everything was fine :-(
> 
> We should have (at least in debug mode) some sort of "leak checker", if tablet data exists that doesn't belong to any table, it's a bug. It would have caught this bug.

We hit an error without this fix, as reported in https://github.com/scylladb/scylladb/issues/17627:
```
ERROR 2024-03-05 00:43:05,896 [shard 0:strm] raft_topology - topology change coordinator fiber got error data_dictionary::no_such_column_family (Can't find a column family with UUID 6c7881d0-da80-11ee-9723-746b0771e35e)
```

We can call on_internal_error_noexcept that will abort when abort_on_internal_error is enabled (in non-release modes)
but otherwise continue, rather than throwing an exception.

nyh commented on cdc/log.cc:
```diff
@@ -216,6 +216,9 @@ class cdc::cdc_service::impl : service::migration_listener::empty_listener {
     void on_before_drop_column_family(const schema& schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        if (schema.is_view()) {
+            return;
+        }
```
Why is this extra check necessary? Won't we just realize that the view does _not_ have a CDC log, and do nothing? Did anything fail without this extra check?

bhalevy replied:
I didn't test specifically with CDC and so I didn't want to change the flow for CDC.
Maybe this is worth at least a FIXME comment and an issue opened about this to consider removing this if statement if everything works without it, but that said, it's legitimate to consider that views do not have CDC logs and short-circuit their handling on this path.

bhalevy replied:
fixed in next version

nyh replied:
What was fixed?

bhalevy replied:
oh, I must have replied on the wrong thread, sorry.
nothing changed here indeed.

bhalevy replied:
removed this condition in v4.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -121,3 +121,49 @@ def test_tablets_are_dropped_when_dropping_table(cql, skip_without_tablets):
+
+def test_tablets_are_dropped_when_dropping_view(cql, skip_without_tablets):
+    ksdef = "WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1'} AND TABLETS = {'initial': 1}"
+    with new_test_keyspace(cql, ksdef) as keyspace:
```
Nitpick:
You don't have to create a new keyspace in all these tests - this creation slows down the test and makes it longer and harder to read (and more indented). You can just use the same "test_keyspace" fixture in all these tests. It will use tablets.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -121,3 +121,49 @@ def test_tablets_are_dropped_when_dropping_table(cql, skip_without_tablets):
+    with new_test_keyspace(cql, ksdef) as keyspace:
+        keyspace_name = keyspace
+        table_name = ""
+        view_name = ""
```
Again all these variable-settings/renaming/copying are unecessary ands just make the test harder to understand.

bhalevy replied:
fixed in next version

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -121,3 +121,49 @@ def test_tablets_are_dropped_when_dropping_table(cql, skip_without_tablets):
+        keyspace_name = keyspace
+        table_name = ""
+        view_name = ""
+        with pytest.raises(InvalidRequest):
```
I had to read this test several times to understand that you expect the new_test_table() function itself to fail, because you created a view in this table and don't remove it, so the table deletion will fail. I think this is not a good use of new_test_table() and you could have just used CREATE TABLE directly (anyway, you lost the protection of deleting the table on exception - the table will be deleted when you delete the keyspace (this test really assumes a new keyspace and not a fixture).

But more importantly, I don't understand why this test needs to exist in the first place -  the DROP TABLE got refused, so nothing should have happened. Testing the tablets didn't get deleted is strange given that nothing at all should have happened (neither the table nor the view changed in any way, didn't get deleted, their tablets didn't, and so on).

bhalevy replied:
> the DROP TABLE got refused, so nothing should have happened.

should => worth verification

bhalevy replied:
fixed in next version by using internal test functions accepting flags for edge-cases
I hope that would be clearer

nyh commented on service/migration_manager.cc:
```diff
@@ -726,7 +727,8 @@ future<std::vector<mutation>> prepare_column_family_update_announcement(storage_
     try {
         auto& db = sp.local_db();
-        auto&& old_schema = db.find_column_family(cfm->ks_name(), cfm->cf_name()).schema(); // FIXME: Should we lookup by id?
+        const auto& cf = db.find_column_family(cfm->id());
```
I see you fixed the FIXME while at it :-) I don't remember why we had this FIXME instead of just doing it...

bhalevy replied:
me neither.  I can open-code find_column_family here and lookup the table_id by `(ks_name, cf_name)`.
I don't think it can be different than `cfm->id()`.  Id that case, it would mean a new version of the table, not of its schema (i.e. after drop and recreate with the same name).
Maybe historically we may have thought the id would denote the schema version?

bhalevy replied:
left as is in next version

nyh commented on service/migration_manager.cc:
```diff
@@ -746,7 +748,24 @@ future<std::vector<mutation>> prepare_column_family_update_announcement(storage_
+                const index_metadata& index = old_schema->all_indices().at(name);
+                schema_ptr view;
+                try {
+                    view = db.find_schema(old_schema->ks_name(), name + "_index");
```
Please consider using the index_table_name() function instead of open-coding the `+ "_index"`.

bhalevy replied:
sure

bhalevy replied:
defined one in cql-pytest/utils.py and used it in test_tablets

nyh replied:
Didn't you change this code to use index_table_name()?
Note that my comment was about the C code, **not** about the test...

bhalevy replied:
Fixed in v4

nyh reviewed (changes requested):
Very nice PR. I just left a few minor requests.

bhalevy commented on test/cql-pytest/test_tablets.py:
```diff
@@ -107,3 +107,17 @@ def test_describe_initial_tablets(cql, skip_without_tablets):
+def test_tablets_are_dropped_when_dropping_table(cql, skip_without_tablets):
+    ksdef = "WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1'} AND TABLETS = {'initial': 1}"
+    with new_test_keyspace(cql, ksdef) as keyspace:
+        keyspace_name = keyspace
```
good point

bhalevy replied:
fixed in next version

bhalevy commented on test/cql-pytest/test_tablets.py:
```diff
@@ -107,3 +107,17 @@ def test_describe_initial_tablets(cql, skip_without_tablets):
+    ksdef = "WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1'} AND TABLETS = {'initial': 1}"
+    with new_test_keyspace(cql, ksdef) as keyspace:
+        keyspace_name = keyspace
+        table_name = ""
```
I want it to be outside the scope of the "with ..." block
Maybe Python allows that but it looks so awkward and anti-patternish that I wanted to "declare" the variable first.

bhalevy replied:
fixed in next version

bhalevy commented on test/cql-pytest/test_tablets.py:
```diff
@@ -107,3 +107,17 @@ def test_describe_initial_tablets(cql, skip_without_tablets):
         assert "and tablets = {'initial': 1}" in desc.one().create_statement.lower()
+
+
+def test_tablets_are_dropped_when_dropping_table(cql, skip_without_tablets):
```
fixed in next version

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -127,3 +127,50 @@ def verify_tablets_presence(expected:bool=True):
+    view_pk = "c, pk"
+    cql.execute(f"CREATE MATERIALIZED VIEW {keyspace_name}.{view_name} AS SELECT * FROM {table_name} WHERE {where} PRIMARY KEY ({view_pk})")
+
+    def verify_tablets_presence(table_expected:bool=True, view_expected:bool=True):
```
It seems like you repeat the same function verify_tablets_presence() multiple times. Maybe it makes sense to make it into a outer-scope function (in test_tablets.py, not in util.py!)  instead of an inner function? It can get the table name as a parameter (and in this code, it will be called twice, once for the table and once for the view).

bhalevy replied:
ok, I'll take a stab at it

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -127,3 +127,50 @@ def verify_tablets_presence(expected:bool=True):
     verify_tablets_presence(expected=False)
+
+
+def _test_tablets_are_dropped_when_dropping_table_with_view(cql, keyspace_name, attempt_drop_table:bool=False):
```
Please consider using pytest's "paramterize" instead of this ugly underscore idiom from dtest :-(

bhalevy replied:
I thought parametrize is a dtest idiom not a generic one

nyh replied:
It's the opposite - dtest always used that underscore thing, and parametrize was only recently adopted after dtest moved to pytest. You can see in test/cql-pytest quite a few examples of parametrize. It's useful for exactly this case - when you want to run the same test once with true and once with false, or some other case of several runs. pytest also makes it easy to run just one of the parameterizations (e.g., `test/cql-pytest/run somefile.py::sometest[dog]` runs the "dog" case) and they are nicely reported in the output and counted as separate tests.

bhalevy replied:
used in v3

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -174,3 +174,38 @@ def test_tablets_are_dropped_when_dropping_view(cql, test_keyspace, skip_without
     _test_tablets_are_dropped_when_dropping_table_with_view(cql, test_keyspace, attempt_drop_table=True)
+
+
+def _test_tablets_are_dropped_when_dropping_index(cql, keyspace_name:str, drop_index:bool):
```
Again, please use "paramterize" instead of this underscore idiom.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -174,3 +174,38 @@ def test_tablets_are_dropped_when_dropping_view(cql, test_keyspace, skip_without
+    schema = "pk int PRIMARY KEY, c int"
+    cql.execute(f"CREATE TABLE {keyspace_name}.{table_name} ({schema})")
+    index_name = unique_name()
+    cql.execute(f"CREATE INDEX {index_name} ON {keyspace_name}.{table_name} (c)")
```
Nitpick: I'm starting to be sorry I asked you not to create a keyspace. If your test uses a common keyspace, it should be careful not to leave behind tables etc. in this keyspace when it fails. And this test isn't careful about that.

But I don't insist. We don't expect the test to fail anyway, and it's not like the DynamoDB tests where we actually pay with real money for every table that we leave behind :-)

nyh replied:
It's not very important. In Alternator it's very important (leaving behind test tables on AWS DynamoDB costs us money) but in CQL, if you don't leave many of them behind (and you won't, unless many tests fail), it's not a big problem.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -1148,6 +1148,7 @@ class tablet_allocator_impl : public tablet_allocator::impl
         auto rs = abstract_replication_strategy::create_replication_strategy(ksm.strategy_name(), params);
         if (auto&& tablet_rs = rs->maybe_as_tablet_aware()) {
             auto tm = _db.get_shared_token_metadata().get();
+            lblogger.info("Creating tablets for {}.{} id={}", s.ks_name(), s.cf_name(), s.id());
```
This isn't interesting for users.

bhalevy replied:
ok, will demote to debug level

bhalevy replied:
demoted in v3

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -1158,6 +1159,7 @@ class tablet_allocator_impl : public tablet_allocator::impl
         auto&& rs = ks.get_replication_strategy();
         if (rs.uses_tablets()) {
             auto tm = _db.get_shared_token_metadata().get();
+            lblogger.info("Dropping tablets for {}.{} id={}", s.ks_name(), s.cf_name(), s.id());
```
Or this.

bhalevy replied:
ditto

bhalevy replied:
demoted in v3

avikivity commented on service/migration_manager.cc:
```diff
@@ -746,7 +748,24 @@ future<std::vector<mutation>> prepare_column_family_update_announcement(storage_
+                    mlogger.error("Could not find schema for index {}.{} id={}", old_schema->ks_name(), name + "_index", index.id());
+                    continue;
+                }
+                db.get_notifier().before_drop_column_family(*view, mutations, ts);
```
Do we have more asymmetries between tables and views?

bhalevy replied:
you mean views and indices?

nyh replied:
@avikivity what do you mean "asymmetries"?

As I noted in the past, the root of the need for this patch (and commits https://github.com/scylladb/scylladb/commit/88a5ddabcec625434b2df0ed82a63c36245af35f and https://github.com/scylladb/scylladb/commit/8181e287312c2b3eefa222eb6946cdae29195aa5 where I added the tablets for views and indexes) is the idea that you need "callbacks" to create and delete tablets - and you need to remember to call those callbacks in multiple places. Perhaps a more reliable approach would have been to have just one single function that can **create** a table, and the code that created a table, view, or index would all have had to call this single function - and then we would make this single function create the tablets. But this is not what we have today - we need to remember to call the callbacks from any code which can create or delete a table, even if this "table" is a view (or a view backing an index).

avikivity replied:
asymmetry - callback called on table but not on view

nyh replied:
It's impossible to know :-( The "notifier" architecture makes it possible to create a table mutation without calling the notifier. I thought that last time I tracked all the missing places but it turns out I only fixed the view **creation** and now Benny discovered I forgot the view **deletion**. I don't know if there are more places like this. Testing should discover all these bugs (I discovered the missing creation notifiers call in the very first tests that I wrote for views and for secondary index).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7371/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7371/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8501/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/468/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7371/)
#### Build Details:
- Duration: 4 hr 59 min
- Builder: i-04be91214abab0c60 (m5ad.8xlarge)

bhalevy commented:
In v3 ([5fe2036](https://github.com/scylladb/scylladb/commit/5fe2036b625cbf1f332b495802adaee5a2c2d079)):
- rebased
- define standalone `verify_tablets_presence` function in test_tablets.py that is used by subsequent tests
- used `pytest.mark.parametrize` for test variants
- handle errors in test and cleanup by unconditionally dropping view/index/table on exception
- demoted additional log message in tablet_allocator to debug log_level

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -107,3 +107,22 @@ def test_describe_initial_tablets(cql, skip_without_tablets):
+    else:
+        assert not res, f"{keyspace_name}.{table_name} was found in system.tablets after it was dropped"
+
+# Test that when a tablets-enabled table is dropped, all of its tablets are dropped with it.
```
Thanks, I think the test is much clearer now, and thanks for adding a comment. Next time please add a reference to the issue that each test reproduces (since you did write these tests to reproduce a specific issue!), but I won't insist now - this review has dragged on long enough.

bhalevy replied:
Added references to the issue in v4

nyh replied:
I don't see it here?

bhalevy replied:
This particular test, for a table that has no views or indices, has no issue.
It is working without this series.
The next tests reference the issue fixed in this series.

nyh commented on test/cql-pytest/test_tablets.py:
```diff
@@ -126,3 +126,42 @@ def test_tablets_are_dropped_when_dropping_table(cql, test_keyspace, skip_withou
+
+
+# Test that when a view of a tablets-enabled table is dropped, all of its tablets are dropped with it.
+@pytest.mark.parametrize("attempt_drop_table", [False, True])
```
Nitpick (but again, I don't insist, it's not important):
You didn't need to make attempt_drop_table optional (i.e., you didn't need two tests here): As I noted earlier, the "attempt" to drop a table which doesn't succeed has absolutely no effect (and your test verifies that, partially), so you can do it as part of the regular test - you don't need to repeat the regular part of the test twice.

bhalevy replied:
removed `attempt_drop_table` parameter in v4

nyh reviewed (approved):
Thanks. I basically approve this patch (as I said before, it's very good and nicely tested), but I'd like you to take one last look at the comments that are still open. You can choose to reject some of them (and I mentioned that explicitly in some of them), but there in one or two places that you said you accepted my comment and will change it - but it seems you didn't, so please take one more look. Thanks.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7389/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7389/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/476/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8509/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7389/)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> Didn't you change this code to use index_table_name()?
> Note that my comment was about the C code, not about the test...

Oops, I missed that.  I thought you meant the test code, not the C(++) code
Will fix and submit a new version.

bhalevy commented:
In v4 ([3b5f442](https://github.com/scylladb/scylladb/commit/3b5f442081541f007a7165c7fef432eab898d64d)):
- dropped condition from cdc `on_before_drop_column_family` to let it discover that the view/index has no cdc log
- prepare_column_family_update_announcement: use secondary_index::index_table_name instead of string concat
- test_tablets:
  - added references in comments to https://github.com/scylladb/scylladb/issues/17627
  - test_tablets_are_dropped_when_dropping_table_with_view: dropped `attempt_drop_table` parameter

tgrabiec commented on service/migration_manager.cc:
```diff
@@ -746,7 +748,25 @@ future<std::vector<mutation>> prepare_column_family_update_announcement(storage_
             db.get_notifier().before_update_column_family(*cfm, *old_schema, mutations, ts);
+
+            // Notify drop indices
+            auto diff = difference(old_schema->all_indices(), cfm->all_indices());
```
This duplicates logic from make_update_indices_mutations() which is called from make_update_table_mutations(), invoked above. Why not call the notifier there?

bhalevy replied:
First, sending notifications from schema_tables seems like a layering violation.
Second, this is not an exact duplicate.
make_update_indices_mutations calls `cf.get_index_manager().create_view_for_index` which creates a new view schema, while for this purpose we need `db.find_schema` since the index's tablets use that table_id.
We can also look up the view schema in the table's `views()`, but that's a bit less efficient.

tgrabiec replied:
> First, sending notifications from schema_tables seems like a layering violation.

We already call db.get_notifier().before_create_column_family() from there. Note that this is not a generic notification, but an extension point for mutation building. It's not supposed to change the db state, just add more mutations to the result.

> view schema, while for this purpose we need db.find_schema since the index's tablets use that table_id.

We can call find_schema() from there.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7396/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7396/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8516/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7396/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/486/)
#### Build Details:
- Duration: 2 hr 29 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
In v5 ([5bfca73](https://github.com/scylladb/scylladb/commit/5bfca73b303f3bee7fa421a29caebadd34d71a41))
- moved notification about dropped indices when dropping a table with dependent indices to schema_tables make_update_indices_mutations
- improved error handling in test_tablet.py for cleaning up view/table if the test throws
